src/tests/ftest/container/container_redundancy.py

```diff
@@ -0,0 +1,152 @@
+        self.daos_cmd = None
+
+    def create_test_container(self):
+        """Create a container and write objects."""
+        self.container.create()
+        self.container.write_objects(
+            self.inputs.rank.value[0], self.inputs.object_class.value)
+
+    def verify_rank_has_objects(self):
+        """Verify the first rank to be excluded has at least one object."""
+        rank_list = self.container.get_target_rank_lists(" before rebuild")
+        objects = {
+            rank: self.container.get_target_rank_count(rank, rank_list)
+            for rank in self.inputs.rank.value
+        }
+        self.assertGreater(
+            objects[self.inputs.rank.value[0]], 0,
+            "#No objects written to rank {}".format(self.inputs.rank.value[0]))
+
+    def verify_rank_has_no_objects(self):
+        """Verify the excluded rank has zero objects."""
+        rank_list = self.container.get_target_rank_lists(" after rebuild")
+        objects = {
+            rank: self.container.get_target_rank_count(rank, rank_list)
+            for rank in self.inputs.rank.value
+        }
+        for rank in self.inputs.rank.value:
+            self.assertEqual(
+                objects[rank], 0,
+                "#Excluded rank {} still has objects".format(rank))
+
+    def verify_cont_rf_healthstatus(self, expected_rf, expected_health):
+        """Verify the container redundancy factor and health status."""
+        result = self.daos_cmd.container_get_prop(
+                      pool=self.pool.uuid,
+                      cont=self.container.uuid)
+        rf = re.search(r"Redundancy Factor\s*([A-Za-z0-9-]+)",
+                       str(result.stdout)).group(1)
+        health = re.search(r"Health\s*([A-Z]+)", str(result.stdout)).group(1)
+        self.assertEqual(
+            rf, expected_rf,
+            "#Container redundancy factor mismatch, actual: {},"
+            " expected: {}.".format(rf, expected_rf))
+        self.assertEqual(
+            health, expected_health,
+            "#Container health-status mismatch, actual: {},"
+            " expected: {}.".format(health, expected_health))
+
+    def start_rebuild(self):
+        """Start the rebuild process and check for container properties."""
+        self.log.info(
+            "==>(1)Check for container rf and health-status "
+            "before rebuild: HEALTHY")
+        rf = ''.join(self.container.properties.value.split(":"))
+        self.verify_cont_rf_healthstatus(rf, "HEALTHY")
+
+        # Exclude the first rank from the pool to initiate rebuild
+        self.server_managers[0].stop_ranks(
+            [self.inputs.rank.value[0]], self.d_log)
+        self.log.info(
+            "==>(2)Check for container rf and health-status "
+            "after 1st rank rebuild: HEALTHY")
+        self.verify_cont_rf_healthstatus(rf, "HEALTHY")
+        # Wait for rebuild to start
+        self.pool.wait_for_rebuild(True, 1)
+
+    def execute_during_rebuild(self):
```

Comment:
I was talking to @wangdi1 about this test and we think it is better to stop the ranks simultaneously instead of sequentially.
So something like this when stopping ranks:
1. stop ranks 3 and 4 simultaneously
2. wait for rebuild to start
2. query container prop, expect HEALTHY
3. wait for rebuild to complete
4. stop ranks 3, 4, and 5 simultaneously
5. wait for rebuild to start
6. query container prop, expect UNCLEAN
7. wait for rebuild to complete

Reply:
After implemented the above test steps, discovered an issue on step-7, please see DAOS-8126.
We should test container redundancy factor with server-ranks restart sequentially as well as simultaneously.
I will create a new JIRA for the above tests.

Reply:
XueZhao explained the problem in DAOS-8126.
I think one way to do this is to keep the functional positive and negative tests as 2 separate variants/tests where servers are restarted.

1. stop ranks simultaneously (3/4 or 3/4/5)
2. wait for rebuild to start
3. query container prop
4. compare expected container status (HEALTHY or UNCLEAN)

I don't think we should keep the sequential server shutdown in the test. There is no guarantee rank 3 or 4 does not complete rebuild before rank 5 is killed, in which case the testcase will fail.

This testcase is also no doing online IO as described in the ticket DAOS-6270 using ior. I think it's good to check that ior should fail when number of rebuilds exceed rf.

Reply:
Please see new commit for the script and test.yaml update.